File: ios/Classes/PdEventListener.swift
//
//  PdEventListener.swift
//  flutter_pd
//
//  Created by Yuya Matsuo on 2020/09/17.
//

import libpd_ios

class PdEventListener: NSObject, PdListener {
  static var instances: [Int: PdEventListener] = [:]
  static func remove(id: Int) -> PdEventListener? {
    return instances.removeValue(forKey: id)
  }

  private let callback: (Any) -> Void

  init(id: Int, callback: @escaping (Any) -> Void) {
    self.callback = callback
    super.init()
    Self.instances[id] = self
  }

  func receiveMessage(
    _ message: String,
    withArguments arguments: [Any],
    fromSource source: String
  ) {
    callback([
      "type": "message",
      "from": source,
      "symbol": message,
      "args": arguments,
    ])
  }

  func receive(_ received: Float, fromSource source: String) {
    callback([
      "type": "float",
      "from": source,
      "value": received,
    ])
  }

  func receiveSymbol(_ symbol: String, fromSource source: String) {
    callback([
      "type": "symbol",
      "from": source,
      "value": symbol,
    ])
  }

  func receiveBang(fromSource source: String) {
    callback([
      "type": "bang",
      "from": source,
    ])
  }

  func receiveList(_ list: [Any], fromSource source: String) {
    callback([
      "type": "list",
      "from": source,
      "value": list,
    ])
  }
}

File: android/src/main/kotlin/com/niusounds/flutter_pd/PdEventListener.kt
package com.niusounds.flutter_pd

import org.puredata.core.PdListener
import java.lang.ref.WeakReference

class PdEventListener(id: Int,
                      private val callback: (Any) -> Unit) : PdListener {
  companion object {
    private val instances = mutableMapOf<Int, WeakReference<PdEventListener>>()

    fun remove(id: Int): PdEventListener? {
      return instances.remove(id)?.get()
    }
  }

  init {
    instances[id] = WeakReference(this)
  }

  override fun receiveMessage(source: String, symbol: String, vararg args: Any?) {
    callback(mapOf(
        "type" to "message",
        "from" to source,
        "symbol" to symbol,
        "args" to args.asList()
    ))
  }

  override fun receiveFloat(source: String, x: Float) {
    callback(mapOf(
        "type" to "float",
        "from" to source,
        "value" to x
    ))
  }

  override fun receiveSymbol(source: String, symbol: String) {
    callback(mapOf(
        "type" to "symbol",
        "from" to source,
        "value" to symbol
    ))
  }

  override fun receiveBang(source: String) {
    callback(mapOf(
        "type" to "bang",
        "from" to source
    ))
  }

  override fun receiveList(source: String, vararg args: Any?) {
    callback(mapOf(
        "type" to "list",
        "from" to source,
        "value" to args.asList()
    ))
  }
}